src/main/java/org/opentripplanner/routing/algorithm/strategies/PriceBasedRemainingWeightHeuristic.java

```diff
@@ -0,0 +1,95 @@
+
+    private double lat;
+    private double lon;
+    private double priceCostWeight;
+    private double originalCostWeight;
+    private EuclideanRemainingWeightHeuristic originalHeuristic;
+
+    public PriceBasedRemainingWeightHeuristic(Map<CostFunction.CostCategory, Double> costCategoryWeights){
+        priceCostWeight = Optional.ofNullable(costCategoryWeights.get(CostFunction.CostCategory.PRICE_ASSOCIATED)).orElse(DEFAULT_PRICE_COST_WEIGHT);
+        originalCostWeight = Optional.ofNullable(costCategoryWeights.get(CostFunction.CostCategory.ORIGINAL)).orElse(DEFAULT_ORIGINAL_COST_WEIGHT);
+        originalHeuristic = new EuclideanRemainingWeightHeuristic();
+    }
+
+    @Override
+    public void initialize(RoutingRequest request, long abortTime) {
+        Vertex target = request.rctx.target;
+        lat = target.getLat();
+        lon = target.getLon();
+        originalHeuristic.initialize(request, abortTime);
+    }
+
+    @Override
+    public double estimateRemainingWeight(State s) {
+        Vertex currentLocation = s.getVertex();
+        double remainingDistance = SphericalDistanceLibrary.distance(currentLocation.getLat(), currentLocation.getLon(), lat, lon);
+
+        double speed;
+        int remainingTime;
+        double estimatedFuturePrice = 0;
+        Edge backEdge = s.getBackEdge();
+        if (Objects.nonNull(s.getCurrentVehicle()) && Objects.nonNull(backEdge) && backEdge instanceof StreetEdge ) {
+            speed = s.getCurrentVehicle().getMaxSpeedInMetersPerSecond((StreetEdge) backEdge);
+            remainingTime = (int) (remainingDistance / speed);
+            estimatedFuturePrice = chooseBestFuturePrice(s.getCurrentVehicle(),
+                    s.getTimeTraversedInCurrentVehicleInSeconds(), s.getDistanceTraversedInCurrentVehicle(),
+                    remainingTime, remainingDistance);
+        }
+
+        return priceCostWeight * estimatedFuturePrice + originalCostWeight * originalHeuristic.estimateRemainingWeight(s);
+    }
+
+    private double chooseBestFuturePrice(VehicleDescription vehicle,
+                                             int previousTravelTimeInSeconds,
+                                             double previousDistanceInMeters,
+                                             int remainingTravelTimeInSeconds,
+                                             double remainingDistanceInMeters) {
+        VehiclePricingPackage evaluatedPackage = vehicle.getVehiclePricingPackage(0);
+        BigDecimal chosenEstimatedTotalPrice = evaluatedPackage.computeTotalPrice(
+                previousTravelTimeInSeconds + remainingTravelTimeInSeconds,
+                (int) (previousDistanceInMeters + remainingDistanceInMeters));
+        BigDecimal chosenEstimatedPreviousPrice = evaluatedPackage.computeTotalPrice(previousTravelTimeInSeconds, (int) previousDistanceInMeters);
+
+        BigDecimal evaluatedTotalPrice;
+        for (int i = 1; i < vehicle.getVehiclePricingPackages().size(); i++) {
```

Comment:
to chyba powinno być od `0`, ale może lepiej nich to będzie pętla `foreach`.
EDIT: zauważyłem, że pierwsza paczka jest rozpatrzona przed pętlą - w takim razie ja bym przed pętlą ustawił `chosenEstimatedTotalPrice = MAXVALUE` i zrobił pętlę `foreach` po wszystkich elementach tak, żeby na pewno pierwszy element był wybrany jako lepszy od `MAXVALUE`. Wtedy znika kopiowanie kodu ;)